src/vehicle.cpp

```diff
@@ -6466,3 +6475,24 @@ void vehicle::calc_mass_center( bool use_precalc ) const
         mass_center_no_precalc_dirty = false;
     }
 }
+
+bool vehicle::is_floating() const
+{
+    // Pure land vehicle
+    if ( !valid_wheel_config( true ) ) {
+        return false;
+    }
+
+    // Pure boat
+    if ( !valid_wheel_config( false ) ) {
+        return true;
+    }
+
+    // Otherwise, the vehicle is amphibious. We consider it
+    // to be floating if >2/3 of the wheels are submerged
+    if ( g->m.vehicle_wheel_traction( *this ) < 0 ) {
```

Comment:
Negative wheel traction is a hack to sink vehicles. It isn't totally wrong to use it here, but it's not right either.